src/session.rs

```diff
@@ -422,6 +479,14 @@ impl<'l> SessionRunArgs<'l> {
             *tensor = ptr::null_mut();
         }
     }
+
+    fn maybe_reset_run_metadata(&mut self) {
+        if let Some(t) = self.run_metadata.as_ref() {
+            if !t.inner().is_null() {
+                self.run_metadata = unsafe { Some(Buffer::new_unallocated()) };
```

Comment:
This isn't safe, because unallocated buffers are in an unsafe state and shouldn't exist outside of unsafe blocks.  A safer way to do this is to create it inside `Session::run`'s unsafe block and assign it to `SessionRunArgs::run_metadata` after it's been filled.

Reply:
The metadata buffer is now created in `Session::run` and converted to `Vec<u8>` before being stored in `SessionRunArgs::run_metadata`.

With the changes above and the run_metadata buffer becoming a local variable, we either need to create the pointer inside the unsafe block or need a second unsafe block exclusively for the creation of the metadata buffer. I've gone for the latter as it keeps the pointer creation outside the unsafe block. If there's something speaking against it, I'll be happy to go for the other option.